inputremapper/injection/macros/macro.py

```diff
@@ -328,6 +328,27 @@ async def task(handler):
 
         self.tasks.append(task)
 
+    def add_toggle(self, symbol):
+        """Press when calling toggle the first time, release it next time."""
+        _type_check_symbol(symbol)
+
+        async def task(handler):
+            key = f"_toggle_{symbol}"
+
+            resolved_symbol = _resolve(symbol, [str])
+            code = _type_check_symbol(resolved_symbol)
+
+            if macro_variables[key]:
+                macro_variables[key] = False
+                handler(EV_KEY, code, 0)
+                await self._keycode_pause()
+            else:
+                macro_variables[key] = True
+                handler(EV_KEY, code, 1)
+                await self._keycode_pause()
```

Comment:
we can avoid using the global variable dict and all the overhead, and the potential name conflict with user variables:
```suggestion
        def f():
            while True:
                resolved_symbol = _resolve(symbol, [str])
                code = _type_check_symbol(resolved_symbol)
                yield EV_KEY, code, 1
                resolved_symbol = _resolve(symbol, [str])
                code = _type_check_symbol(resolved_symbol)
                yield EV_KEY, code, 0
        toggle = f()

        async def task(handler):
            handler(*next(toggle))
            await self._keycode_pause()

```
I didn't test this yet but it avoids the global dict altogether and the current state is stored inside the generator, which is initialized when parsing the macro.

Reply:
the generator would be unique for each call to `toggle`, wouldn't it?

I think pressing the key twice would correctly toggle it, but `toggle(KEY_A).toggle(KEY_A)` would inject two key-down events as far as I can tell

Reply:
yeah `toggle(KEY_A).toggle(KEY_A)` would send two key down events. and on the second press it would send two key up events.

Reply:
`toggle(KEY_A).toggle(KEY_A)` can also be written like `key_down(KEY_A).key_up(KEY_A)`, not sure if there are any potential other issues. But I am fine with this either way.